From: Joerg Roedel <jroedel@suse.de>
Date: Mon, 11 Jun 2018 16:02:47 +0200
Subject: KVM: x86: Sync back MSR_IA32_SPEC_CTRL to VCPU data structure
Patch-mainline: no, different implementation
References: bsc#1096242, bsc#1096281

Sync MSR_IA32_SPEC_CTRL to vcpu->spec_ctrl to make sure the values are never
going to be out-of-sync.

Signed-off-by: Joerg Roedel <jroedel@suse.de>
Signed-off-by: Jiri Kosina <jkosina@suse.cz>
---
 arch/x86/kernel/cpu/spec_ctrl.c |    1 +
 arch/x86/kvm/svm.c              |    3 +++
 arch/x86/kvm/vmx.c              |    3 +++
 3 files changed, 7 insertions(+)

--- a/arch/x86/kernel/cpu/spec_ctrl.c
+++ b/arch/x86/kernel/cpu/spec_ctrl.c
@@ -66,6 +66,7 @@ void x86_spec_check(void)
 		ibpb_state = 1;
 
 		setup_force_cpu_cap(X86_FEATURE_SPEC_CTRL);
+		setup_force_cpu_cap(X86_FEATURE_MSR_SPEC_CTRL);
 	}
 
 	if (boot_cpu_data.x86_vendor == X86_VENDOR_AMD) {
--- a/arch/x86/kvm/svm.c
+++ b/arch/x86/kvm/svm.c
@@ -4998,6 +4998,9 @@ static void svm_vcpu_run(struct kvm_vcpu
 #endif
 #endif
 
+	if (boot_cpu_has(X86_FEATURE_MSR_SPEC_CTRL))
+		svm->spec_ctrl = native_read_msr(MSR_IA32_SPEC_CTRL);
+
 	x86_spec_ctrl_restore_host(svm->spec_ctrl, svm->virt_spec_ctrl);
 
 	reload_tss(vcpu);
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -8830,6 +8830,9 @@ static void __noclone vmx_vcpu_run(struc
 #endif
 	      );
 
+	if (boot_cpu_has(X86_FEATURE_MSR_SPEC_CTRL))
+		vmx->spec_ctrl = native_read_msr(MSR_IA32_SPEC_CTRL);
+
 	x86_spec_ctrl_restore_host(vmx->spec_ctrl, 0);
 
 	/* Eliminate branch target predictions from guest mode */
